File: store/763.划分字母区间.c
/*
 * @lc app=leetcode.cn id=763 lang=c
 *
 * [763] 划分字母区间
 */

// @lc code=start

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int *partitionLabels(char *s, int *returnSize)
{
  int i, max = -1, top = 0, count = 0, hash[26] = {0}, *res;
  res = (int *)malloc(sizeof(int) * strlen(s));
  for (i = 0; i < strlen(s); i++)
  {
    hash[s[i] - 'a'] = i;
  }
  for (i = 0; i < strlen(s); i++)
  {
    max = max > hash[s[i] - 'a'] ? max : hash[s[i] - 'a'];
    count++;
    if (max == i)
    {
      res[top++] = count;
      count = 0;
    }
  }
  *returnSize = top;
  return res;
}
// @lc code=end

File: store/763.划分字母区间.js
/*
 * @lc app=leetcode.cn id=763 lang=javascript
 *
 * [763] 划分字母区间
 */

// @lc code=start
/**
 * @param {string} s
 * @return {number[]}
 */
var partitionLabels = function (s) {
  let map = new Map(),
    res = [],
    maxRight = 0;
  for (let i = s.length - 1; i > -1; i--) {
    if (!map.has(s[i])) map.set(s[i], i);
  }
  for (let i = 0; i < s.length; i++) {
    maxRight = maxRight > map.get(s[i]) ? maxRight : map.get(s[i]);
    if (i === maxRight) res.push(i + 1);
  }
  return res.map((val, index) => {
    if (index > 0) return val - res[index - 1];
    return val;
  });
};
// @lc code=end
